### PR TITLE
PointCulling Trait and Iterator cleanup

### PR DESCRIPTION
**Pull request by catevita**
The iterator are splitted on 
-iterator that retrieve nodes
-iterator that scan points on a single node

The PointLocation in which the Points that have to be retrieved has been streamlined,
the methods contains and intersects are part of the trait PointCulling.
This works for an Aabb3, Obb, OrientedBeam and Frustum

### CODE REVIEW

**Comment by catevita** [comment]
still quite a big PR, I tried my best to remove some other additional changes
PTAL @nnmm @feuerste

File: src/math.rs [hunk 1]
```diff
@@ -276,19 +301,43 @@ impl<S: BaseFloat> Obb<S> {
     }
 }
 
+impl<S> PointCulling<S> for Obb<S>
```
Comment (nnmm) [inline src/math.rs]: Nit: Why was this moved?

File: src/octree/batch_iterator.rs [hunk 2]
```diff
@@ -1,31 +1,47 @@
...
 /// size for batch
 pub const NUM_POINTS_PER_BATCH: usize = 500_000;
 
-///possible kind of iterators that can be evaluated in batch of points in BatchIterator
 #[allow(clippy::large_enum_variant)]
-#[derive(Clone)]
-pub enum PointCulling {
-    Any(),
+#[derive(Debug, Clone)]
+pub enum GeoFence {
```
Comment (nnmm) [inline src/octree/batch_iterator.rs]: I think it might be clearer if we name this PointLocation, and what's currently called PointLocation can become PointQuery. GeoFence has a specific meaning: https://en.wikipedia.org/wiki/Geo-fence

File: src/octree/batch_iterator.rs [hunk 3]
```diff
@@ -151,16 +148,27 @@ impl<'a> BatchIterator<'a> {
...
-        let mut iterator: Box<Iterator<Item = Point>> = match &self.culling {
-            PointCulling::Any() => Box::new(self.octree.all_points()),
-            PointCulling::Aabb(aabb) => Box::new(self.octree.points_in_box(aabb)),
-            PointCulling::Obb(obb) => Box::new(self.octree.points_in_obb(obb)),
-            PointCulling::Frustum(frustum) => Box::new(self.octree.points_in_frustum(frustum)),
-            PointCulling::OrientedBeam(beam) => Box::new(self.octree.points_in_oriented_beam(beam)),
-        };
-        iterator.try_for_each(|point: Point| point_stream.push_point_and_callback(point))?;
-        point_stream.callback()
+        //todo (catevita) mutable function parallelization
```
Comment (nnmm) [inline src/octree/batch_iterator.rs]: I'm not sure I understand and/or we need this. In the end, the user calls try_for_each_batch. Are you proposing the callbacks should happen on several threads in parallel? I think that's not super useful, since it requires synchronization on the user side again. Better to (as a user) e.g. request points in non-overlapping boxes from several threads.

File: src/octree/mod.rs [hunk 4]
```diff
@@ -302,50 +300,35 @@ impl Octree {
...
+        let container = location.get_point_culling();
+        let filter_func = move |node_id: &NodeId, octree: &Octree| -> bool {
+            let current = &octree.nodes[&node_id];
+            container.intersects(&current.bounding_cube.to_aabb3())
+        };
+        NodeIdsIterator::new(&self, filter_func)
     }
 
-    pub fn points_in_frustum<'a>(
+    /// Returns the ids of all nodes that cut or are fully contained in 'aabb'.
```
Comment (nnmm) [inline src/octree/mod.rs]: Nit: Let's be consistent with two vs three slashes

File: src/math.rs [hunk 5]
```diff
@@ -16,11 +16,59 @@ use cgmath::{
...
+use std::convert::From;
+use std::fmt::Debug;
 use std::ops::Mul;
 
+pub trait PointCulling<S>: objekt::Clone + Debug + Sync + Send
+where
+    S: BaseFloat + Sync + Send,
+{
+    fn contains(&self, point: &Point3<S>) -> bool;
+    fn intersects(&self, aabb: &Aabb3<S>) -> bool;
```
Comment (nnmm) [inline src/math.rs]: Could you add a TODO here to return Relation?

File: src/math.rs [hunk 6]
```diff
@@ -16,11 +16,59 @@ use cgmath::{
...
+    fn transform(&self, isometry: &Isometry3<S>) -> Box<PointCulling<S>>;
```
Comment (nnmm) [inline src/math.rs]: This can probably be impl PointCulling, I'd think.

File: src/octree/mod.rs [hunk 7]
```diff
@@ -397,3 +380,54 @@ fn maybe_push_node(
         });
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct Frustum<S: BaseFloat> {
+    pub matrix: Matrix4<S>,
```
Comment (nnmm) [inline src/octree/mod.rs]: Could you remind me why this contains the second matrix? I forgot. Also, the fields probably do not need to be public.

File: src/octree/batch_iterator.rs (hunk 2 above)
Comment (catevita) [reply src/octree/batch_iterator.rs]: good idea, I was running short on names!

File: src/octree/batch_iterator.rs (hunk 3 above)
Comment (catevita) [reply src/octree/batch_iterator.rs]: As told offline, this PR only takes care of implementing the trait, the parallelization follows up next

File: src/octree/batch_iterator.rs (hunk 2 above)
Comment (catevita) [reply src/octree/batch_iterator.rs]: done!

File: src/math.rs (hunk 1 above)
Comment (catevita) [reply src/math.rs]: the methods have been moved to implement the trait - I hope to understand the question

File: src/octree/mod.rs (hunk 4 above)
Comment (catevita) [reply src/octree/mod.rs]: this was meant to be a minimal doc comment for the method points_in_node, do we use doc comment at all?

File: src/octree/mod.rs (hunk 7 above)
Comment (catevita) [reply src/octree/mod.rs]: i moved it in math, the matrix is used by the contains function

// TODO(ksavinash9) update after https://github.com/rustgd/collision-rs/issues/101 is resolved.
pub fn contains<S: BaseFloat>(projection_matrix: &Matrix4<S>, point: &Point3<S>) -> bool {

the issue in collision is not resolved yet

File: src/math.rs (hunk 5 above)
Comment (catevita) [reply src/math.rs]: done

File: src/math.rs (hunk 6 above)
Comment (catevita) [reply src/math.rs]: Unfortunately this does not work inside a trait, from [rfcs](https://github.com/rust-lang/rfcs/blob/master/text/1522-conservative-impl-trait.md)

_"impl Trait may only be written within the return type of a freestanding or inherent-impl function, not in trait definitions or any non-return type position. "_

File: src/math.rs (hunk 5 above)
Comment (feuerste) [reply src/math.rs]: Nit: Our formatting convention for TODOs is:
```rust
// TODO(username): Do something.
```
Can you please adjust it everywhere in this PR?

File: point_cloud_client/src/bin/test.rs [hunk 8]
```diff
@@ -54,8 +53,8 @@ fn main() {
     let num_points = args.num_points;
     let point_cloud_client = PointCloudClient::new(&args.locations, OctreeFactory::new())
         .expect("Couldn't create octree client.");
-    let point_location = PointLocation {
-        culling: PointCulling::Aabb(Aabb3::new(args.min, args.max)),
```
Comment (feuerste) [inline point_cloud_client/src/bin/test.rs]: Any specific reason to remove the min/max restriction?
Then the whole point of having min/max arguments goes away...

File: point_viewer_grpc/src/bin/octree_benchmark.rs [hunk 9]
```diff
@@ -71,14 +72,20 @@ fn server_benchmark(octree_directory: &Path, num_points: u64) {
...
-            println!("Streamed {}M points", counter / 1_000_000);
-        }
-        counter += 1;
-        if counter == num_points {
+    let all_points = PointQuery {
+        location: PointLocation::AllPoints(),
+        global_from_local: None,
+    };
+    let mut batch_iterator = BatchIterator::new(&octree, &all_points, 1_000_000);
+    let mut num_batches = 0;
```
Comment (feuerste) [inline point_viewer_grpc/src/bin/octree_benchmark.rs]: Nit: How about having a const variable `BATCH_SIZE`, removing `num_batches`, and just working on `counter`?

File: point_viewer_grpc/src/service.rs [hunk 10]
```diff
@@ -145,12 +137,11 @@ impl proto_grpc::Octree for OctreeService {
...
-            OctreeQuery::Frustum(frustum_matrix),
-            &req.octree_id,
-            &ctx,
-            resp,
-        )
+        let point_location = PointQuery {
+            location: octree::PointLocation::Frustum(frustum_matrix),
+            global_from_local: None,
+        };
+        self.stream_points_back_to_sink(point_location, &req.octree_id, &ctx, resp)
```
Comment (feuerste) [inline point_viewer_grpc/src/service.rs]: It seems `global_from_local` is `None` whenever calling `self.stream_points_back_to_sink`. How about just passing a `location` and constructing the `PointQuery` inside `stream_points_back_to_sink`?

File: point_viewer_grpc/src/service.rs [hunk 11]
```diff
@@ -266,57 +264,67 @@ impl OctreeService {
...
+                            v
+                        })
+                        .collect();
+
+                    reply.colors = match p_data.layers.get(&"color".to_string()) {
+                        Some(LayerData::U8Vec4(data)) => data
+                            .iter()
+                            .map(|p| {
+                                let mut v = point_viewer::proto::Color::new();
+                                v.set_red(f32::from(p.x) / 255.);
```
Comment (feuerste) [inline point_viewer_grpc/src/service.rs]: How about using `Color::to_f32` to perform the conversion?

File: src/math.rs [hunk 12]
```diff
@@ -13,14 +13,63 @@
...
+where
+    S: BaseFloat + Sync + Send,
+{
+    fn contains(&self, point: &Point3<S>) -> bool;
+    // todo (catevita) return Relation
+    fn intersects(&self, aabb: &Aabb3<S>) -> bool;
+    fn transform(&self, isometry: &Isometry3<S>) -> Box<PointCulling<S>>;
+}
+
+// PointCulling for Aabb3
```
Comment (feuerste) [inline src/math.rs]: Nit: remove comment, this is clear

File: src/math.rs [hunk 13]
```diff
@@ -150,22 +199,6 @@ impl<'a, S: BaseFloat> Mul<&'a Vector3<S>> for &'a Isometry3<S> {
     }
 }
 
-impl<S: BaseFloat> Into<Decomposed<Vector3<S>, Quaternion<S>>> for Isometry3<S> {
```
Comment (feuerste) [inline src/math.rs]: Any reason why the generics were removed here and instead having an `f64` implementation only?

File: src/math.rs [hunk 14]
```diff
@@ -297,27 +347,12 @@ impl OrientedBeam {
...
-        isometry: &Isometry3<f64>,
-        half_extent: &Vector2<f64>,
-    ) -> [Point3<f64>; 4] {
-        let corner_from = |x: f64, y: f64| {
-            isometry.rotation.rotate_point(Point3::new(x, y, 0.0)) + isometry.translation
+    fn precompute_corners(isometry: &Isometry3<S>, half_extent: &Vector2<S>) -> [Point3<S>; 4] {
+        let corner_from = |x, y| {
+            isometry
+                .rotation
+                .rotate_point(Point3::new(x, y, S::from(0.0).unwrap()))
```
Comment (feuerste) [inline src/math.rs]: Nit: Use `S::zero()` instead of `S:from(0.0)`

File: src/math.rs [hunk 15]
```diff
@@ -428,3 +486,54 @@ mod tests {
...
+    fn intersects(&self, aabb: &Aabb3<S>) -> bool {
+        match self.frustum.contains(aabb) {
+            Relation::Cross => true,
+            Relation::In => true,
+            Relation::Out => false,
+        }
+    }
+
+    fn transform(&self, isometry: &Isometry3<S>) -> Box<PointCulling<S>> {
+        let matrix = Matrix4::from(Decomposed {
```
Comment (feuerste) [inline src/math.rs]: Here we could nicely make use of the conversion methods between `Decomposed` and `Isometry3`, which were deleted above...

File: src/octree/batch_iterator.rs [hunk 16]
```diff
@@ -151,16 +148,27 @@ impl<'a> BatchIterator<'a> {
...
+            .clone()
+            .map(|t| t.inverse());
+        let mut point_stream = PointStream::new(self.batch_size, local_from_global, &mut func);
+        // nodes iterator: retrieve nodes
+        let node_id_iterator = self.octree.nodes_in_location(self.point_location);
+        // operate on nodes
+        for node_id in node_id_iterator {
+            let point_iterator = self.octree.points_in_node(self.point_location, node_id);
+            for point in point_iterator {
+                if let Err(err) = point_stream.push_point_and_callback(point) {
```
Comment (feuerste) [inline src/octree/batch_iterator.rs]: Can we just do `point_stream.push_point_and_callback(point)?;` instead?

File: src/math.rs (hunk 13 above)
Comment (catevita) [reply src/math.rs]: this there is indeed some automatic or manual merging gone wrong! I am restoring them

File: src/math.rs (hunk 14 above)
Comment (catevita) [reply src/math.rs]: done

File: src/math.rs (hunk 12 above)
Comment (catevita) [reply src/math.rs]: done

File: src/math.rs (hunk 5 above)
Comment (catevita) [reply src/math.rs]: done

File: point_viewer_grpc/src/bin/octree_benchmark.rs (hunk 9 above)
Comment (catevita) [reply point_viewer_grpc/src/bin/octree_benchmark.rs]: done

File: point_cloud_client/src/bin/test.rs (hunk 8 above)
Comment (catevita) [reply point_cloud_client/src/bin/test.rs]: done

Comment (catevita) [reply point_cloud_client/src/bin/test.rs]: i think it also was a copy-paste error

File: point_viewer_grpc/src/service.rs (hunk 10 above)
Comment (catevita) [reply point_viewer_grpc/src/service.rs]: if it is ok for you i'd like to leave this pieces of code as is, as the grpc server is not intensively used, so if we have to support query with local/global transforms the PointQuery struct is already prepared to support that.

**Comment by catevita** [comment]
I did some benchmarking running ` time (./target/release/octree_benchmark ~/octree_marco/ --no-client)` on my workstation. Results are quite aligned, typically these

new:
`real	0m3.662s
user	0m3.500s
sys	0m0.160s`

old
`real	0m3.563s
user	0m3.408s
sys	0m0.152s`

File: src/math.rs (hunk 15 above)
Comment (catevita) [reply src/math.rs]: done

File: src/octree/batch_iterator.rs (hunk 16 above)
Comment (catevita) [reply src/octree/batch_iterator.rs]: done!

File: point_viewer_grpc/src/bin/octree_benchmark.rs (hunk 9 above)
Comment (catevita) [reply point_viewer_grpc/src/bin/octree_benchmark.rs]: done

File: point_viewer_grpc/src/service.rs (hunk 11 above)
Comment (catevita) [reply point_viewer_grpc/src/service.rs]: I changed it, it adds some lines but the performance seems not to be affected from that

**Comment by feuerste** [comment]
Can you please remove the `.DS_Store` file?

File: src/math.rs [hunk 17]
```diff
@@ -13,14 +13,61 @@
...
+{
+    fn contains(&self, point: &Point3<S>) -> bool;
+    // TODO (catevita) return Relation
+    fn intersects(&self, aabb: &Aabb3<S>) -> bool;
+    fn transform(&self, isometry: &Isometry3<S>) -> Box<PointCulling<S>>;
+}
+
+impl<S> PointCulling<S> for Aabb3<S>
+where
+    S: 'static + BaseFloat + Sync + Send,
```
Comment (feuerste) [inline src/math.rs]: What do we `'static` need for?

File: src/math.rs (hunk 5 above)
Comment (feuerste) [reply src/math.rs]: Supernit: Still not in the format
```
// TODO(username): Do something.
```
everywhere.

File: src/math.rs (hunk 13 above)
Comment (feuerste) [reply src/math.rs]: It would be clearer to move them back exactly where they were before, so they don't show up as diff...

File: src/math.rs (hunk 5 above)
Comment (catevita) [reply src/math.rs]: My bad, I just paid attention to the caps! now it should be finally fixed!

**Comment by catevita** [comment]
> Can you please remove the `.DS_Store` file?

Indeed it shows up in the list, but it as removed file

**Comment by feuerste** [comment]
> > Can you please remove the `.DS_Store` file?
> 
> Indeed it shows up in the list, but it as removed file

Ah, I see, thanks, I overlooked this...

**Comment by catevita** [comment]
@wally-the-cartographer merge

**Comment by catevita** [comment]
@wally-the-cartographer merge

**Comment by feuerste** [comment]
@wally-the-cartographer merge

**Comment by catevita** [comment]
@gaschler sorry to bother you, Wally seems to be down again, could you help me merging this branch or pointing to the right person that mantains Wally?

**Comment by catevita** [comment]
thanks!
